address/address.c

```diff
@@ -44,8 +44,22 @@ const char AddressSpecials[] = "@.,:;<>[]\\\"()";
 /**
  * is_special - Is this character special to an email address?
  * @param ch Character
+ * @param mask Bitmask of characters 32-95 that are special (others are always zero)
  */
-#define is_special(ch) strchr(AddressSpecials, ch)
+#define is_special(ch, mask) ((ch) >= 32 && (ch) < 96 && ((mask >> ((ch) - 32)) & 1))
+
+/* AddressSpecials, for is_special() */
+#define ADDRESS_SPECIAL_MASK 0x380000015c005304ULL
+
+/* AddressSpecials except " ( . \ */
+#define USER_SPECIAL_MASK 0x280000015c001200ULL
+
+/* AddressSpecials except ( . [ \ ] */
+#define DOMAIN_SPECIAL_MASK 0x000000015c001204ULL
+
+/* AddressSpecials except ( , . [ \ ] */
+#define ROUTE_SPECIAL_MASK 0x000000015c000204ULL
+
```

Comment:
Can you please add instructions how one builds such a mask?

Reply:
How about:
```
/**
 * Character bitmasks
 *
 * The following four bitmasks are used for matching characters at speed, instead of `strchr(3)`.
 *
 * To generate them, consider the value of each character, e.g. `,` == 44.
 * Now set the 44th bit of a `unsigned long long`.  Repeat for each character.
 *
 * To test a character, say `(` (40), we check if the 40th bit it set in the mask.
 *
 * The characters we want to test, AddressSpecials, range in value between 34 and 93.
 * This is too large for a integer type, so we subtract 32 to bring the values
 * down to 2 to 61.
 */
```

Reply:
Done. I added a few extra notes and fixed a typo here and there.